contracts/instruments/StakedPut.sol

```diff
@@ -0,0 +1,326 @@
+            underlying,
+            strikeAsset,
+            collateralAsset,
+            expiry,
+            currentPrice,
+            optionType,
+            inputToken
+        );
+
+        premium = adapter.premium(optionTerms, wbtcSize);
+        totalCost = amt + premium;
+    }
+
+    function buyInstrument(BuyInstrumentParams calldata params) public payable
+    {
+        require(params.paymentToken == ethAddress, 'input must be eth');
+        buyLpFromAdapter(params.paymentToken, params.lpAmt, params.exchangeName, params.tradeAmt, params.minWbtcAmtOut,params.minDiggAmtOut);
+
+        uint256 positionID = buyPutFromAdapter(params);
+
+        emit PositionCreated(
+            msg.sender,
+            positionID,
+            params.exchangeName,
+            params.lpAmt
+        );
+    }
+
+    function buyLpFromAdapter(address tokenInput, uint256 amt, string memory exchangeName, uint256 tradeAmt, uint256 minWbtcAmtOut, uint256 minDiggAmtOut) public payable
+    {
+        iUniswapAdapter.delegateBuyLp(tokenInput, amt, exchangeName, tradeAmt, minWbtcAmtOut, minDiggAmtOut);
+    }
+
+    function exercisePosition(uint256 positionID)
+        external
+        nonReentrant
+        returns (uint256 totalProfit)
+    {
+        InstrumentPosition storage position =
+            instrumentPositions[msg.sender][positionID];
+        require(!position.exercised, "Already exercised");
+
+        uint256 strikePrice;
+        uint32 optionID;
+
+        strikePrice = position.putStrikePrice;
+        optionID = position.putOptionID;
+
+        uint256 amount = position.amount;
+
+        uint256 profit =
+            adapter.delegateExerciseProfit(
+                optionsAddress,
+                optionID,
+                amount
+            );
+        if (profit > 0) {
+            adapter.delegateExercise(
+                optionsAddress,
+                optionID,
+                amount,
+                msg.sender
+            );
+        }
+
+        totalProfit += profit;
+
+        position.exercised = true;
+
+        emit Exercised(msg.sender, positionID, totalProfit);
+    }
+
+    function exerciseProfit(address account, uint256 positionID)
+        external
+        view
+        returns (uint256)
+    {
+        InstrumentPosition storage position =
+            instrumentPositions[account][positionID];
+
+        if (position.exercised) return 0;
+
+        uint256 profit = 0;
+
+        uint256 amount = position.amount;
+
+        uint256 strikePrice;
+        uint32 optionID;
+
+        strikePrice = position.putStrikePrice;
+        optionID = position.putOptionID;
+
+        address adapterAddress = factory.getAdapter(venue);
+        require(adapterAddress != address(0), "Adapter does not exist");
+
+        bool exercisable = adapter.canExercise(optionsAddress, optionID, amount);
+        if (!exercisable) {
+            return 0;
+        }

```

Comment:
We can remove this check entirely because we assume the adapters return `0` when the option cannot be exercised